### PR TITLE
feat: Users creation on user management dashboard.

### PR DESCRIPTION
**Pull request by Chakravarthy7102**
## Summary of change

Adds API's to create Passwordless and Emailpassword recipe users from user management dashboard.

## Related issues

-   Link to issue1 here
-   Link to issue1 here

## Test Plan

(Write your test plan here. If you changed any code, please provide us with clear instructions on how you verified your changes work. Bonus points for screenshots and videos!)

## Documentation changes

(If relevant, please create a PR in our [docs repo](https://github.com/supertokens/docs), or create a checklist here highlighting the necessary changes)

## Checklist for important updates

-   [x] Changelog has been updated
-   [ ] `coreDriverInterfaceSupported.json` file has been updated (if needed)
    -   Along with the associated array in `lib/ts/version.ts`
-   [ ] `frontendDriverInterfaceSupported.json` file has been updated (if needed)
-   [x] Changes to the version if needed
    -   In `package.json`
    -   In `package-lock.json`
    -   In `lib/ts/version.ts`
-   [x] Had run `npm run build-pretty`
-   [x] Had installed and ran the pre-commit hook
-   [ ] If new thirdparty provider is added,
    -   [ ] update switch statement in `recipe/thirdparty/providers/configUtils.ts` file, `createProvider` function.
    -   [ ] add an icon on the user management dashboard.
-   [x] Issue this PR against the latest non released version branch.
    -   To know which one it is, run find the latest released tag (`git tag`) in the format `vX.Y.Z`, and then find the latest branch (`git branch --all`) whose `X.Y` is greater than the latest released tag.
    -   If no such branch exists, then create one from the latest released branch.
-   [ ] If have added a new web framework, update the `add-ts-no-check.js` file to include that
-   [ ] If added a new recipe / api interface, then make sure that the implementation of it uses NON arrow functions only (like `someFunc: function () {..}`).
-   [ ] If added a new recipe, then make sure to expose it inside the recipe folder present in the root of this repo. We also need to expose its types.

## Remaining TODOs for this PR

-   [ ] Item1
-   [ ] Item2

### CODE REVIEW

File: lib/ts/recipe/dashboard/constants.ts [hunk 1]
```diff
@@ -34,6 +34,9 @@ export const USERROLES_ROLE_API = "/api/userroles/role";
 export const USERROLES_PERMISSIONS_API = "/api/userroles/role/permissions";
 export const USERROLES_REMOVE_PERMISSIONS_API = "/api/userroles/role/permissions/remove";
 
+export const CREATE_EMAIL_PASSWORD_USER = "/api/user/create/emailpassword";
+export const CREATE_PASSWORDLESS_USER = "/api/user/create/passwordless";
```
Comment (rishabhpoddar) [inline lib/ts/recipe/dashboard/constants.ts:38]:
```suggestion
export const CREATE_EMAIL_PASSWORD_USER = "/api/user/emailpassword";
export const CREATE_PASSWORDLESS_USER = "/api/user/passwordless";
```

File: lib/ts/recipe/dashboard/api/listTenants.ts [hunk 2]
```diff
@@ -43,6 +48,12 @@ export default async function listTenants(
 ): Promise<Response> {
     let tenantsRes = await Multitenancy.listAllTenants(userContext);
     let finalTenants: TenantListTenantType[] = [];
+    let passwordlessContactMethod: PasswordlessContactMethod | undefined = undefined;
+
+    try {
+        const passwordlessRecipe = PasswordlessRecipe.getInstanceOrThrowError();
```
Comment (rishabhpoddar) [inline lib/ts/recipe/dashboard/api/listTenants.ts:54]: user may have also init thirdpartypasswordless.. that also needs to be taken into account.

File: lib/ts/recipe/dashboard/api/user/create/emailpasswordUser.ts [hunk 3]
```diff
@@ -0,0 +1,53 @@
...
+export const createEmailPasswordUser = async (
+    _: APIInterface,
+    tenantId: string,
+    options: APIOptions,
+    __: any
+): Promise<Response> => {
+    try {
+        EmailPasswordRecipe.getInstanceOrThrowError();
+    } catch (error) {
+        return {
```
Comment (rishabhpoddar) [inline lib/ts/recipe/dashboard/api/user/create/emailpasswordUser.ts:27]: user may have init thirdpartyemailpassword..

File: lib/ts/recipe/dashboard/api/listTenants.ts [hunk 4]
```diff
@@ -43,6 +48,12 @@ export default async function listTenants(
 ): Promise<Response> {
     let tenantsRes = await Multitenancy.listAllTenants(userContext);
     let finalTenants: TenantListTenantType[] = [];
```
Comment (rishabhpoddar) [inline lib/ts/recipe/dashboard/api/listTenants.ts:50]: the public tenant will have all three login methods as enabled by default.. which means in the UI, you will see email password and passwordless methods to create users, even if the user has not init those recipes!

File: lib/ts/recipe/dashboard/api/user/create/emailpasswordUser.ts [hunk 5]
```diff
@@ -0,0 +1,53 @@
...
+    }
+
+    if (password === undefined || typeof password !== "string") {
+        throw new STError({
+            message: "Required parameter 'password' is missing or has an invalid type",
+            type: STError.BAD_INPUT_ERROR,
+        });
+    }
+
+    const response = await EmailPassword.signUp(tenantId, email, password);
```
Comment (rishabhpoddar) [inline lib/ts/recipe/dashboard/api/user/create/emailpasswordUser.ts:50]: You need to run the email and password validator.. and if that doesnt match, you need to send input validation error. This also requires changes in the dashboard api spec

File: lib/ts/recipe/dashboard/api/user/create/passwordlessUser.ts [hunk 6]
```diff
@@ -0,0 +1,95 @@
...
+export const createPasswordlessUser = async (
+    _: APIInterface,
+    tenantId: string,
+    options: APIOptions,
+    __: any
+): Promise<Response> => {
+    let passwordlessRecipe: PasswordlessRecipe | undefined = undefined;
+
+    try {
+        passwordlessRecipe = PasswordlessRecipe.getInstanceOrThrowError();
```
Comment (rishabhpoddar) [inline lib/ts/recipe/dashboard/api/user/create/passwordlessUser.ts:33]: user may have init thirdpartypasswordless as well

File: lib/ts/recipe/dashboard/api/user/create/passwordlessUser.ts [hunk 7]
```diff
@@ -0,0 +1,95 @@
...
+import Passwordless from "../../../../passwordless";
+import PasswordlessRecipe from "../../../../passwordless/recipe";
+import { User } from "../../../../../types";
+import RecipeUserId from "../../../../../recipeUserId";
+import { parsePhoneNumber } from "libphonenumber-js/max";
+
+type Response =
+    | {
+          status: string;
+          createdNewRecipeUser: boolean;
```
Comment (rishabhpoddar) [inline lib/ts/recipe/dashboard/api/user/create/passwordlessUser.ts:12]: what will the frontend show in case (createdNewRecipeUser === false || (createdNewRecipeUser === true && user.loginMethods.length === 1))? It means it didn't create a new user (sign in) or account linking happened.

Same goes for emailpassword user creation too

**Review by rishabhpoddar — CHANGES_REQUESTED** [review]

File: lib/ts/recipe/dashboard/constants.ts (hunk 1 above)
Comment (Chakravarthy7102) [reply lib/ts/recipe/dashboard/constants.ts:38]: done

File: lib/ts/recipe/dashboard/api/user/create/passwordlessUser.ts (hunk 7 above)
Comment (Chakravarthy7102) [reply lib/ts/recipe/dashboard/api/user/create/passwordlessUser.ts:12]: it says the user with this email/phone already exists.

Comment (Chakravarthy7102) [reply lib/ts/recipe/dashboard/api/user/create/passwordlessUser.ts:12]: when createdNewRecipeUser === false on frontend we show a message say `user already exists` and when the accounts are linked we open the primary user details page in new tab.

File: lib/ts/recipe/dashboard/api/user/create/emailpasswordUser.ts (hunk 5 above)
Comment (Chakravarthy7102) [reply lib/ts/recipe/dashboard/api/user/create/emailpasswordUser.ts:50]: done

File: lib/ts/recipe/dashboard/constants.ts (hunk 1 above)
Comment (Chakravarthy7102) [reply lib/ts/recipe/dashboard/constants.ts:38]: TODO: need to change the folder structure as well. But no doing right now as moving files from a folder will create a new files and it would not be easy to review the changes.

File: lib/ts/recipe/dashboard/api/user/create/passwordlessUser.ts (hunk 6 above)
Comment (Chakravarthy7102) [reply lib/ts/recipe/dashboard/api/user/create/passwordlessUser.ts:33]: done

File: lib/ts/recipe/dashboard/api/user/create/emailpasswordUser.ts (hunk 3 above)
Comment (Chakravarthy7102) [reply lib/ts/recipe/dashboard/api/user/create/emailpasswordUser.ts:27]: done

File: lib/ts/recipe/dashboard/api/listTenants.ts (hunk 2 above)
Comment (Chakravarthy7102) [reply lib/ts/recipe/dashboard/api/listTenants.ts:54]: done

File: lib/ts/recipe/dashboard/api/listTenants.ts (hunk 4 above)
Comment (Chakravarthy7102) [reply lib/ts/recipe/dashboard/api/listTenants.ts:54]: done

File: lib/ts/recipe/dashboard/api/listTenants.ts [hunk 8]
```diff
@@ -56,6 +76,18 @@ export default async function listTenants(
             passwordless: currentTenant.passwordless,
             thirdParty: currentTenant.thirdParty,
         };
+
+        if (passwordlessContactMethod !== undefined) {
```
Comment (rishabhpoddar) [inline lib/ts/recipe/dashboard/api/listTenants.ts:80]: rewrite this logic.. first get the enabled login methods based on the supertokens.init on the backend sdk, and then further filter that with whats enabled for the tenant. This should apply to all tenants and not just the public tenant

File: lib/ts/recipe/dashboard/api/user/create/emailpasswordUser.ts [hunk 9]
```diff
@@ -0,0 +1,82 @@
...
+    }
+
+    const validatePasswordError = await defaultPasswordValidator(password);
+
+    if (validatePasswordError !== undefined) {
+        return {
+            status: "INPUT_VALIDATION_ERROR",
+            message: validatePasswordError,
+        };
+    }
```
Comment (rishabhpoddar) [inline lib/ts/recipe/dashboard/api/user/create/emailpasswordUser.ts:77]: this is wrong.. you need to run the validators from the user's configured ones (which defaults to the defaultValidators...)

File: lib/ts/recipe/dashboard/api/user/create/emailpasswordUser.ts [hunk 10]
```diff
@@ -0,0 +1,82 @@
...
+            type: STError.BAD_INPUT_ERROR,
+        });
+    }
+
+    const validateEmailError = await defaultEmailValidator(email);
+
+    if (validateEmailError !== undefined) {
+        return {
+            status: "INPUT_VALIDATION_ERROR",
+            message: validateEmailError,
```
Comment (rishabhpoddar) [inline lib/ts/recipe/dashboard/api/user/create/emailpasswordUser.ts:66]: In the UI, how is this error shown? Cause the status is the same as for invalid password and there is no clear way to tell if it's invalid email or password.

File: lib/ts/recipe/dashboard/api/user/create/passwordlessUser.ts [hunk 11]
```diff
@@ -0,0 +1,95 @@
...
+        email !== undefined &&
+        (passwordlessRecipe.config.contactMethod === "EMAIL" ||
+            passwordlessRecipe.config.contactMethod === "EMAIL_OR_PHONE")
+    ) {
+        email = email.trim();
+        const validateError = await defaultValidateEmail(email);
+        if (validateError !== undefined) {
+            return {
+                status: "INPUT_VALIDATION_ERROR",
+                message: validateError,
```
Comment (rishabhpoddar) [inline lib/ts/recipe/dashboard/api/user/create/passwordlessUser.ts:68]: how will the frontend know if it's email or phone validation error?

File: lib/ts/recipe/dashboard/api/user/create/passwordlessUser.ts [hunk 12]
```diff
@@ -0,0 +1,95 @@
...
+        });
+    }
+
+    if (
+        email !== undefined &&
+        (passwordlessRecipe.config.contactMethod === "EMAIL" ||
+            passwordlessRecipe.config.contactMethod === "EMAIL_OR_PHONE")
+    ) {
+        email = email.trim();
+        const validateError = await defaultValidateEmail(email);
```
Comment (rishabhpoddar) [inline lib/ts/recipe/dashboard/api/user/create/passwordlessUser.ts:64]: should use user defined validator (which defaults to defaultValidateEmail)

File: lib/ts/recipe/dashboard/api/user/create/passwordlessUser.ts [hunk 13]
```diff
@@ -0,0 +1,95 @@
...
+            };
+        }
+    }
+
+    if (
+        phoneNumber !== undefined &&
+        (passwordlessRecipe.config.contactMethod === "PHONE" ||
+            passwordlessRecipe.config.contactMethod === "EMAIL_OR_PHONE")
+    ) {
+        const validateError = await defaultValidatePhoneNumber(phoneNumber);
```
Comment (rishabhpoddar) [inline lib/ts/recipe/dashboard/api/user/create/passwordlessUser.ts:78]: should use user defined validator (which defaults to defaultValidatePhoneNumber)

File: lib/ts/recipe/dashboard/api/user/create/passwordlessUser.ts [hunk 14]
```diff
@@ -0,0 +1,95 @@
...
+        if (validateError !== undefined) {
+            return {
+                status: "INPUT_VALIDATION_ERROR",
+                message: validateError,
+            };
+        }
+
+        const parsedPhoneNumber = parsePhoneNumber(phoneNumber);
```
Comment (rishabhpoddar) [inline lib/ts/recipe/dashboard/api/user/create/passwordlessUser.ts:86]: what if parsedPhoneNumber is undefined? See https://github.com/supertokens/supertokens-node/blob/master/lib/ts/recipe/passwordless/api/createCode.ts#L85

File: lib/ts/recipe/dashboard/api/user/create/passwordlessUser.ts [hunk 15]
```diff
@@ -0,0 +1,95 @@
...
+        phoneNumber = parsedPhoneNumber.format("E.164");
+    }
+
+    const response = await Passwordless.signInUp(
```
Comment (rishabhpoddar) [inline lib/ts/recipe/dashboard/api/user/create/passwordlessUser.ts:90]: what if passwordless is not defined.. you have checked for that right above..

File: lib/ts/recipe/dashboard/recipe.ts [hunk 16]
```diff
@@ -79,7 +79,7 @@ import getRolesForUser from "./api/userroles/getRolesForUser";
 import removeUserRole from "./api/userroles/removeUserRole";
 import createRoleOrAddPermissions from "./api/userroles/roles/createRoleOrAddPermissions";
 import { createEmailPasswordUser } from "./api/user/create/emailpasswordUser";
-import { createPasswordlessUser } from "./api/user/create/passwordlessUser";
+import { createPasswordlessUser } from "./api/user//create/passwordlessUser";
```
Comment (rishabhpoddar) [inline lib/ts/recipe/dashboard/recipe.ts:82]: why 2 `//`?

**Review by rishabhpoddar — CHANGES_REQUESTED** [review]

File: lib/ts/recipe/dashboard/recipe.ts (hunk 16 above)
Comment (Chakravarthy7102) [reply lib/ts/recipe/dashboard/recipe.ts]: removed

File: lib/ts/recipe/dashboard/api/user/create/passwordlessUser.ts (hunk 15 above)
Comment (Chakravarthy7102) [reply lib/ts/recipe/dashboard/api/user/create/passwordlessUser.ts]: done

File: lib/ts/recipe/dashboard/api/user/create/passwordlessUser.ts (hunk 14 above)
Comment (Chakravarthy7102) [reply lib/ts/recipe/dashboard/api/user/create/passwordlessUser.ts:113]: done

File: lib/ts/recipe/dashboard/api/user/create/passwordlessUser.ts (hunk 13 above)
Comment (Chakravarthy7102) [reply lib/ts/recipe/dashboard/api/user/create/passwordlessUser.ts]: done

File: lib/ts/recipe/dashboard/api/user/create/passwordlessUser.ts (hunk 12 above)
Comment (Chakravarthy7102) [reply lib/ts/recipe/dashboard/api/user/create/passwordlessUser.ts]: done

File: lib/ts/recipe/dashboard/api/user/create/passwordlessUser.ts (hunk 11 above)
Comment (Chakravarthy7102) [reply lib/ts/recipe/dashboard/api/user/create/passwordlessUser.ts]: added separate statuses for email and phone validation errors.

File: lib/ts/recipe/dashboard/api/user/create/emailpasswordUser.ts (hunk 10 above)
Comment (Chakravarthy7102) [reply lib/ts/recipe/dashboard/api/user/create/emailpasswordUser.ts]: added separate statuses for email and password validation errors.

File: lib/ts/recipe/dashboard/api/user/create/emailpasswordUser.ts (hunk 9 above)
Comment (Chakravarthy7102) [reply lib/ts/recipe/dashboard/api/user/create/emailpasswordUser.ts]: done

File: lib/ts/recipe/dashboard/api/listTenants.ts (hunk 8 above)
Comment (Chakravarthy7102) [reply lib/ts/recipe/dashboard/api/listTenants.ts]: done

File: lib/ts/recipe/dashboard/api/getTenantLoginMethodsInfo.ts [hunk 17]
```diff
@@ -0,0 +1,133 @@
...
+    emailPassword: {
+        enabled: boolean;
+    };
+    passwordless: {
+        enabled: boolean;
+        contactMethod?: PasswordlessContactMethod;
+    };
+    thirdParty: {
+        enabled: boolean;
+        providers: ProviderConfig[];
```
Comment (rishabhpoddar) [inline lib/ts/recipe/dashboard/api/getTenantLoginMethodsInfo.ts:38]: why is this required?

File: lib/ts/recipe/dashboard/api/getTenantLoginMethodsInfo.ts [hunk 18]
```diff
@@ -0,0 +1,133 @@
...
+    tenants: TenantLoginMethodType[];
+};
+
+export default async function getTenantLoginMethodsInfo(
+    _: APIInterface,
+    __: string,
+    ___: APIOptions,
+    userContext: any
+): Promise<Response> {
+    const tenantsRes = await Multitenancy.listAllTenants(userContext);
```
Comment (rishabhpoddar) [inline lib/ts/recipe/dashboard/api/getTenantLoginMethodsInfo.ts:53]: why are you calling list all tenants? There is a function to get the tenant info for just a specific tenant id. Why arent you using that?

Comment (rishabhpoddar) [reply lib/ts/recipe/dashboard/api/getTenantLoginMethodsInfo.ts:53]: This API needs to be per tenant right?

File: lib/ts/recipe/dashboard/api/user/create/emailpasswordUser.ts [hunk 19]
```diff
@@ -58,25 +66,49 @@ export const createEmailPasswordUser = async (
...
+        if (validateEmailError !== undefined) {
+            return {
+                status: "EMAIL_VALIDATION_ERROR",
+                message: validateEmailError,
+            };
+        }
+    } else {
+        // this should never happen.
+        throw new Error("emailFormFiled is undefined");
     }
```
Comment (rishabhpoddar) [inline lib/ts/recipe/dashboard/api/user/create/emailpasswordUser.ts:85]: so why is this block there? Just do `emailFormField!. validate(...)`

File: lib/ts/recipe/dashboard/api/user/create/passwordlessUser.ts [hunk 20]
```diff
@@ -57,39 +65,71 @@ export const createPasswordlessUser = async (
...
-        if (validateError !== undefined) {
+        let validationError: string | undefined = undefined;
+
+        if (passwordlessOrThirdpartyPasswordlessRecipe.config.validateEmailAddress !== undefined) {
+            validationError = await passwordlessOrThirdpartyPasswordlessRecipe.config.validateEmailAddress(
+                email,
+                tenantId
+            );
+        } else {
+            validationError = await defaultValidateEmail(email);
```
Comment (rishabhpoddar) [inline lib/ts/recipe/dashboard/api/user/create/passwordlessUser.ts:80]: this is wrong.. if validateEmailAddress us undefined, then the contact method was phone number, so it should never come here. So you can use `passwordlessOrThirdpartyPasswordlessRecipe.config.validateEmailAddress!(...)`

File: lib/ts/recipe/dashboard/api/user/create/passwordlessUser.ts [hunk 21]
```diff
@@ -57,39 +65,71 @@ export const createPasswordlessUser = async (
...
-        if (validateError !== undefined) {
+        let validationError: string | undefined = undefined;
+
+        if (passwordlessOrThirdpartyPasswordlessRecipe.config.validatePhoneNumber !== undefined) {
+            validationError = await passwordlessOrThirdpartyPasswordlessRecipe.config.validatePhoneNumber(
+                phoneNumber,
+                tenantId
+            );
+        } else {
+            validationError = await defaultValidatePhoneNumber(phoneNumber);
```
Comment (rishabhpoddar) [inline lib/ts/recipe/dashboard/api/user/create/passwordlessUser.ts:103]: this is wrong.. no need for this cause if validatePhoneNumber is undefined, then the contact method is email

File: lib/ts/recipe/dashboard/constants.ts [hunk 22]
```diff
@@ -37,6 +37,7 @@ export const USERROLES_REMOVE_PERMISSIONS_API = "/api/userroles/role/permissions
 export const CREATE_EMAIL_PASSWORD_USER = "/api/user/emailpassword";
 export const CREATE_PASSWORDLESS_USER = "/api/user/passwordless";
 
+export const LIST_TENANT_LOGIN_METHODS = "/api/tenants/login-methods";
```
Comment (rishabhpoddar) [inline lib/ts/recipe/dashboard/constants.ts:40]: why is this for all tenants and not just the current tenant? It should be just for the currently selected tenant.

**Review by rishabhpoddar — CHANGES_REQUESTED** [review]

File: lib/ts/recipe/dashboard/api/getTenantLoginMethodsInfo.ts (hunk 17 above)
Comment (Chakravarthy7102) [reply lib/ts/recipe/dashboard/api/getTenantLoginMethodsInfo.ts:38]: removed

File: lib/ts/recipe/dashboard/api/getTenantLoginMethodsInfo.ts (hunk 18 above)
Comment (Chakravarthy7102) [reply lib/ts/recipe/dashboard/api/getTenantLoginMethodsInfo.ts:53]: Addressed this issue on slack in our discussion.

File: lib/ts/recipe/dashboard/api/user/create/emailpasswordUser.ts (hunk 19 above)
Comment (Chakravarthy7102) [reply lib/ts/recipe/dashboard/api/user/create/emailpasswordUser.ts:85]: done

File: lib/ts/recipe/dashboard/api/user/create/passwordlessUser.ts (hunk 20 above)
Comment (Chakravarthy7102) [reply lib/ts/recipe/dashboard/api/user/create/passwordlessUser.ts]: done

File: lib/ts/recipe/dashboard/api/user/create/passwordlessUser.ts (hunk 21 above)
Comment (Chakravarthy7102) [reply lib/ts/recipe/dashboard/api/user/create/passwordlessUser.ts]: done

File: lib/ts/recipe/dashboard/api/user/create/passwordlessUser.ts [hunk 23]
```diff
@@ -0,0 +1,121 @@
...
+    __: any
+): Promise<Response> => {
+    let passwordlessOrThirdpartyPasswordlessRecipe: PasswordlessRecipe | undefined = undefined;
+
+    try {
+        passwordlessOrThirdpartyPasswordlessRecipe = PasswordlessRecipe.getInstanceOrThrowError();
+    } catch (_) {
+        try {
+            passwordlessOrThirdpartyPasswordlessRecipe = ThirdPartyPasswordlessRecipe.getInstanceOrThrowError()
+                .passwordlessRecipe;
```
Comment (Chakravarthy7102) [inline lib/ts/recipe/dashboard/api/user/create/passwordlessUser.ts:43]: Using `passwordlessRecipe` property in the `ThirdpartyPasswordlessRecipe` since if the user didn't defined the validator function in the config then the functions are `undefined`  even if they have defined validators those will be passed down to this `passwordlessRecipe` property according to these [line](https://github.com/supertokens/supertokens-node/blob/master/lib/ts/recipe/thirdpartyemailpassword/recipe.ts#L128) of code.

File: lib/ts/recipe/dashboard/api/user/create/passwordlessUser.ts [hunk 24]
```diff
@@ -0,0 +1,117 @@
...
+    options: APIOptions,
+    __: any
+): Promise<Response> => {
+    let passwordlessRecipe: PasswordlessRecipe | undefined = undefined;
+
+    try {
+        passwordlessRecipe = PasswordlessRecipe.getInstanceOrThrowError();
+    } catch (_) {
+        try {
+            passwordlessRecipe = ThirdPartyPasswordlessRecipe.getInstanceOrThrowError().passwordlessRecipe;
```
Comment (Chakravarthy7102) [inline lib/ts/recipe/dashboard/api/user/create/passwordlessUser.ts:42]: Using the `passwordlessRecipe` property within the `ThirdpartyPasswordlessRecipe` is important. If the user hasn't defined the validator function in the configuration in `ThirdpartyPasswordless`, these validator functions become undefined. However, even if the user has defined validators, they will be passed down to the passwordlessRecipe property. This behavior is detailed in the [code here](https://github.com/supertokens/supertokens-node/blob/master/lib/ts/recipe/thirdpartyemailpassword/recipe.ts#L128)

File: lib/ts/recipe/thirdpartypasswordless/types.ts [hunk 25]
```diff
@@ -56,6 +56,10 @@ export type TypeInput = (
     | {
           contactMethod: "EMAIL_OR_PHONE";
           validateEmailAddress?: (email: string, tenantId: string) => Promise<string | undefined> | string | undefined;
+          validatePhoneNumber?: (
```
Comment (Chakravarthy7102) [inline lib/ts/recipe/thirdpartypasswordless/types.ts:59]: Added missing type

Comment (rishabhpoddar) [inline lib/ts/recipe/thirdpartypasswordless/types.ts:59]: add a test for this

**Review by rishabhpoddar — CHANGES_REQUESTED** [review]

File: lib/ts/recipe/thirdpartypasswordless/types.ts (hunk 25 above)
Comment (Chakravarthy7102) [reply lib/ts/recipe/thirdpartypasswordless/types.ts:59]: done

**Review by rishabhpoddar — APPROVED** [review]

File: lib/ts/recipe/dashboard/api/getTenantLoginMethodsInfo.ts [hunk 26]
```diff
@@ -27,9 +27,11 @@ type TenantLoginMethodType = {
     tenantId: string;
     emailPassword: {
         enabled: boolean;
+        isThirdParty?: boolean;
```
Comment (rishabhpoddar) [inline lib/ts/recipe/dashboard/api/getTenantLoginMethodsInfo.ts:30]: this is a bad name... have another object called thirdPartyEmailPasssword: {enabled: boolean}. Same goes for thirdpartypasswordless

**Review by rishabhpoddar — CHANGES_REQUESTED** [review]

File: lib/ts/recipe/dashboard/api/getTenantLoginMethodsInfo.ts [hunk 27]
```diff
@@ -97,11 +106,9 @@ function normaliseTenantLoginMethodsWithInitConfig(
     if (tenantDetailsFromCore.passwordless.enabled === true) {
         try {
             const thirdpartyPasswordlessRecipe = ThirdPartyPasswordlessRecipe.getInstanceOrThrowError();
-            normalisedTenantLoginMethodsInfo.passwordless.enabled = true;
-            normalisedTenantLoginMethodsInfo.passwordless.contactMethod =
+            normalisedTenantLoginMethodsInfo.thirdPartyPasswordless!.enabled = true;
+            normalisedTenantLoginMethodsInfo.thirdPartyPasswordless!.contactMethod =
```
Comment (rishabhpoddar) [inline lib/ts/recipe/dashboard/api/getTenantLoginMethodsInfo.ts:110]: why do you need to have the `!` at the end of `thirdPartyPasswordless`?

File: lib/ts/recipe/dashboard/api/getTenantLoginMethodsInfo.ts [hunk 28]
```diff
@@ -0,0 +1,141 @@
...
+            const thirdpartyPasswordlessRecipe = ThirdPartyPasswordlessRecipe.getInstanceOrThrowError();
+            normalisedTenantLoginMethodsInfo.thirdPartyPasswordless!.enabled = true;
+            normalisedTenantLoginMethodsInfo.thirdPartyPasswordless!.contactMethod =
+                thirdpartyPasswordlessRecipe.config.contactMethod;
+        } catch (_) {
+            try {
+                const passwordlessRecipe = PasswordlessRecipe.getInstanceOrThrowError();
+                normalisedTenantLoginMethodsInfo.passwordless.enabled = true;
+                normalisedTenantLoginMethodsInfo.passwordless.contactMethod = passwordlessRecipe.config.contactMethod;
+            } catch (_) {}
```
Comment (rishabhpoddar) [inline lib/ts/recipe/dashboard/api/getTenantLoginMethodsInfo.ts:117]: this should go in its own block... users can theoritically have both passwordless and ThirdPartyPasswordlessRecipe enabled!

File: lib/ts/recipe/dashboard/api/getTenantLoginMethodsInfo.ts [hunk 29]
```diff
@@ -0,0 +1,141 @@
...
+
+    if (tenantDetailsFromCore.emailPassword.enabled === true) {
+        try {
+            ThirdPartyEmailPasswordRecipe.getInstanceOrThrowError();
+            normalisedTenantLoginMethodsInfo.thirdPartyEmailPasssword!.enabled = true;
+        } catch (_) {
+            try {
+                EmailPasswordRecipe.getInstanceOrThrowError();
+                normalisedTenantLoginMethodsInfo.emailPassword.enabled = true;
+            } catch (_) {}
```
Comment (rishabhpoddar) [inline lib/ts/recipe/dashboard/api/getTenantLoginMethodsInfo.ts:129]: this should go in its own block... users can theoritically have both EmailPassword and ThirdPartyEmailPasswordRecipe enabled!

**Review by rishabhpoddar — CHANGES_REQUESTED** [review]

File: lib/ts/recipe/dashboard/api/getTenantLoginMethodsInfo.ts (hunk 29 above)
Comment (Chakravarthy7102) [reply lib/ts/recipe/dashboard/api/getTenantLoginMethodsInfo.ts:129]: done

File: lib/ts/recipe/dashboard/api/getTenantLoginMethodsInfo.ts (hunk 28 above)
Comment (Chakravarthy7102) [reply lib/ts/recipe/dashboard/api/getTenantLoginMethodsInfo.ts:117]: done

File: lib/ts/recipe/dashboard/api/getTenantLoginMethodsInfo.ts (hunk 27 above)
Comment (Chakravarthy7102) [reply lib/ts/recipe/dashboard/api/getTenantLoginMethodsInfo.ts:110]: this is because of bad response type that I've chosen before. Now I've updated with the correct set of types..

File: lib/ts/recipe/dashboard/api/getTenantLoginMethodsInfo.ts [hunk 30]
```diff
@@ -104,30 +109,34 @@ function normaliseTenantLoginMethodsWithInitConfig(
...
 
     if (tenantDetailsFromCore.passwordless.enabled === true) {
+        try {
+            const passwordlessRecipe = PasswordlessRecipe.getInstanceOrThrowError();
+            normalisedTenantLoginMethodsInfo.passwordless.enabled = true;
+            normalisedTenantLoginMethodsInfo.passwordless.contactMethod = passwordlessRecipe.config.contactMethod;
+        } catch (_) {}
+    }
+
+    if (tenantDetailsFromCore.thirdParty.enabled === true && tenantDetailsFromCore.passwordless.enabled === true) {
```
Comment (rishabhpoddar) [inline lib/ts/recipe/dashboard/api/getTenantLoginMethodsInfo.ts:119]:
```suggestion
    if (tenantDetailsFromCore.thirdParty.enabled === true || tenantDetailsFromCore.passwordless.enabled === true) {
```

File: lib/ts/recipe/dashboard/api/getTenantLoginMethodsInfo.ts [hunk 31]
```diff
@@ -104,30 +109,34 @@ function normaliseTenantLoginMethodsWithInitConfig(
...
     }
 
     if (tenantDetailsFromCore.emailPassword.enabled === true) {
+        try {
+            EmailPasswordRecipe.getInstanceOrThrowError();
+            normalisedTenantLoginMethodsInfo.emailPassword.enabled = true;
+        } catch (_) {}
+    }
+
+    if (tenantDetailsFromCore.thirdParty.enabled === true && tenantDetailsFromCore.emailPassword.enabled === true) {
```
Comment (rishabhpoddar) [inline lib/ts/recipe/dashboard/api/getTenantLoginMethodsInfo.ts:135]:
```suggestion
    if (tenantDetailsFromCore.thirdParty.enabled === true || tenantDetailsFromCore.emailPassword.enabled === true) {
```

**Review by rishabhpoddar — CHANGES_REQUESTED** [review]

File: lib/ts/recipe/dashboard/api/getTenantLoginMethodsInfo.ts (hunk 31 above)
Comment (Chakravarthy7102) [reply lib/ts/recipe/dashboard/api/getTenantLoginMethodsInfo.ts]: done

File: lib/ts/recipe/dashboard/api/getTenantLoginMethodsInfo.ts (hunk 30 above)
Comment (Chakravarthy7102) [reply lib/ts/recipe/dashboard/api/getTenantLoginMethodsInfo.ts]: done

**Review by rishabhpoddar — APPROVED** [review]
